KAFKA-15361: Process and persist dir info with broker registration

soarez opened the pull request:
Controllers should process and persist directory information from the broker registration request


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soarez commented:
This builds on #14820  - KAFKA-15886: Always specify directories for new partition registrations – so this is marked as draft until #14820 is merged.

**Reviews**: please focus on the second commit, titled "[KAFKA-15361: Process and persist dir info with broker registration](https://github.com/apache/kafka/pull/14838/commits/aa945193cc705df6e6e904b5ab8a9ecfeca4a38b)"

soarez commented:
@cmccabe @pprovenzano @rondagostino PTAL

pprovenzano commented on server-common/src/main/java/org/apache/kafka/common/DirectoryId.java:
```diff
@@ -121,8 +122,38 @@ public static Map<Integer, Uuid> createAssignmentMap(int[] replicas, Uuid[] dire
+     * @return                 true if the directory is considered online, false otherwise
+     */
+    public static boolean isOnline(Uuid dir, List<Uuid> sortedOnlineDirs) {
+        if (UNASSIGNED.equals(dir) || MIGRATING.equals(dir)) {
```
We should only allow for a directory ID to be MIGRATING if we are in migration mode.

soarez replied:
Can you clarify what you mean with "migration mode"?

If you mean zk->kraft migration, then I don't think so. Metadata records may have been persisted in an older version, without directory assignment. That may happen outside of a ZK->KRaft migration.
When such records are found, the default directory for every replica is assumed to be `MIGRATING`.

See in #14820 :
 * `new PartitionRegistration(PartitionRecord record)`
 * `partitionRegistration.merge(PartitionChangeRecord record)`
 * `partitionRegistration.toRecord(...)`
 * `partitionChangeBuilder.setAssignmentChanges()`

cmccabe replied:
> If you mean zk->kraft migration, then I don't think so. Metadata records may have been persisted in an older version, without directory assignment

yes, agreed. @pprovenzano , `MIGRATING` state can happen outside ZK migration.

Honestly, I think `MIGRATING` isn't a great name. Someone who sticks on metadata version 15 won't ever "migrate" to having directory IDs. Maybe calling this `UNKNOWN` would be more descriptive? But that's another discussion I guess :)

soarez replied:
I don't think it's a bad name, we just need to make it clear the migration refers to the record version, and not to the cluster migration. 

You might not remember, but [the name was your suggestion](https://lists.apache.org/thread/38sbj42ssoksjq2p488vdk1992dw3j1t). 😄

cmccabe replied:
Heh. :)

Yeah I'm on the fence. With good enough docs maybe the name is ok...

soarez commented:
@cmccabe @rondagostino @pprovenzano please take a look

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/BrokerHeartbeatManager.java:
```diff
@@ -61,6 +67,11 @@ static class BrokerHeartbeatState {
+        /**
+         * The directories which the broker currently has available.
+         */
+        private final List<Uuid> directories;
```
Can we rename this to `directoriesSorted` or `sortedDirectories` to indicate that this list must be sorted?  We depend on it being sorted when invoking `DirectoryId.isOnline()`.  And when I look at where we invoke that method, it seems that the field `BrokerRegistration.directories` should get a similar rename.

The other alternative is to remove the constraint that the list be sorted within `DirectoryId.isOnline()`.  Since we don't expect to have a large number of directories, I would lean toward this option in order to avoid bugs down the road if we forget to sort a list somewhere.  Could we then remove the sorting of the list in this class and within `BrokerRegistration` and in any tests?  Or am I missing something and we really do need to keep these sorted?

soarez replied:
The point was to use binary search for lookups, see `DirectoryId.isOnline`. Since we don't expect a large list of directories, it may be ok to always to a O(N) search. We can come back to this in the future if it proves to be an issue. I'll change it to avoid any sorting.

cmccabe replied:
Can you just make it sorted? The nondeterminism of unsorted just sucks for testing / debugging / etc.

soarez replied:
Reverted, back to a sorted list, but now clearly named as such.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -284,7 +286,7 @@ ReplicaPlacer replicaPlacer() {
     public void activate() {
         heartbeatManager = new BrokerHeartbeatManager(logContext, time, sessionTimeoutNs);
         for (BrokerRegistration registration : brokerRegistrations.values()) {
-            heartbeatManager.touch(registration.id(), registration.fenced(), -1);
+            heartbeatManager.register(registration.id(), registration.fenced(), registration.directories());
```
This change seems a bit odd to me.  I think it will work, and in fact I think it will do exactly what was being done before, but you moved the creation of the broker entry in the heartbeat manager's internal map into the `register()` method and removed it from the `touch()` method -- which I think is a good change.  But now this reads weirdly because we aren't registering a broker -- the broker should already have been registered by this point, and we just want to "touch" it.  I think this change can be reverted?

soarez replied:
This is in `clusterControlManager.activate()`, a couple of lines above we reset the heartbeatManager with `heartbeatManager = new BrokerHeartbeatManager(logContext, time, sessionTimeoutNs);`. So no broker will be registered yet with that fresh manager.

`register()` here means _register the broker in the heartbeat manager_, not _register the broker with the controller quorum_. 

If we revert this line change we'll trigger `throw new IllegalStateException("Broker " + brokerId + " is not registered.");` in `brokerHeartbeatManager.touch()`.

rondagostino replied:
Ah, okay, yes, makes sense.  Thanks.

cmccabe replied:
I don't see why we need to handle directories in `HeartbeatManager`. I made this comment in the other PR too. HBM is about timing out brokers. It's not about JBOD stuff really.

Directories are hard state so they should be kept in `ClusterControlManager`. And we need to issue a new `BrokerRegistrationRecord` when they change.

soarez replied:
> And we need to issue a new BrokerRegistrationRecord when they change.
We do that in #14902 which builds on this.

I have removed directory state from the heartbeat manager, we now rely only on `ClusterControlManager`, and any persisted metadata state.

rondagostino commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -354,6 +356,25 @@ public ControllerResult<BrokerRegistrationReply> registerBroker(
+        if (featureControl.metadataVersion().isDirectoryAssignmentSupported()) {
+            Set<Uuid> set = new HashSet<>(request.logDirs());
+            if (set.stream().anyMatch(DirectoryId::reserved)) {
+                throw new DuplicateBrokerRegistrationException("Reserved directory ID in request");
```
Is `DuplicateBrokerRegistrationException` appropriate, do you think?  Seems like `BrokerIdNotRegisteredException` might be better, and then we would have to make an update in `testRegisterWithDuplicateDirectoryId()` to check for that other exception class.

soarez replied:
Good point. `BrokerIdNotRegisteredException` is better indeed.

cmccabe replied:
Hmm. `DuplicateBrokerRegistrationException` is certainly the wrong thing, but I don't think `BrokerIdNotRegisteredException` is quite right either. That exception was intended for when you made an RPC that required broker N to be registered (like heartbeat) but broker N was not in fact registered. It wasn't intended for an error returned from the registration RPC itself. After all, it provides no information. OK the broker is not registered, but why?

I can see that there are some lines above which are using `BrokerIdNotRegisteredException` as a generic "didn't register" error, but that doesn't seem quite right...
```
        if (request.isMigratingZkBroker() && !zkRegistrationAllowed()) {
            throw new BrokerIdNotRegisteredException("Controller does not support registering ZK brokers.");
        }

        if (!request.isMigratingZkBroker() && featureControl.inPreMigrationMode()) {
            throw new BrokerIdNotRegisteredException("Controller is in pre-migration mode and cannot register KRaft " +
                "brokers until the metadata migration is complete.");
        }
```

The "pre-migration" thing is just a straightforward logic error that should never happen. Maybe `RuntimeException` (and consequent controller failover) is OK for that.

The `request.isMigratingZkBroker()` check probably deserves its own error code since it's a weird scenario that should have been given its own error code by the migration KIP. Not sure if it's worth doing now but... probably

Your thing is more like INVALID_REQUEST, since someone is sending obvious nonsense which superficially resembles the schema.

soarez replied:
Updated all three cases to throw `InvalidRequestException` instead.

rondagostino commented on metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:
```diff
@@ -347,7 +347,7 @@ private void maybePublishMetadata(MetadataDelta delta, MetadataImage image, Load
         metrics.updateLastAppliedImageProvenance(image.provenance());
         metrics.setCurrentMetadataVersion(image.features().metadataVersion());
-        if (uninitializedPublishers.isEmpty()) {
+        if (!uninitializedPublishers.isEmpty()) {
```
Nice catch

cmccabe replied:
Sigh. Looks like this was introduced in KAFKA-14538, which I reviewed. It really only bites in scenarios where you're running a pre-NO_OP record MV, which is rare these days.

Thanks for the fix.

rondagostino commented on metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java:
```diff
@@ -272,8 +273,16 @@ public void testRegistrationWithIncorrectClusterId() {
+        return Stream.of(
+                MetadataVersion.IBP_3_3_IV2,
+                MetadataVersion.IBP_3_3_IV3,
+                MetadataVersion.latest()
```
Maybe we want to add in `MetadataVersion.LATEST_PRODUCTION` as well?

soarez replied:
Added

rondagostino reviewed (commented):
Looks pretty good, just a few small things.

rondagostino commented:
Looks like there are lots of test failures, too.  Seems like you will have to look (e.g. https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/9/tests/) and figure out what is going on.

pprovenzano reviewed (approved):
LGTM

rondagostino reviewed (approved):
LGTM assuming tests pass.  Thanks for the PR and adjustments along the way!

soarez commented:
Thanks for the review.

@rondagostino AFAICT, the tests are passing on all JDK versions. Did I miss something?

rondagostino commented:
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/13/tests
```
208 tests have failed
There are 20 new tests failing, 188 existing failing and 272 skipped
```
@soarez it seems that something is not correct.

cmccabe commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -354,6 +357,25 @@ public ControllerResult<BrokerRegistrationReply> registerBroker(
                 "brokers until the metadata migration is complete.");
         }
+
+        if (featureControl.metadataVersion().isDirectoryAssignmentSupported()) {
```
It should not be legal to send an empty `logDirs` array when metadataVersion >= 3.7-IV2

Since setting that MV on the controller requires that all the brokers support it.

It is, of course, OK to send logDirs when MV < 3.7-IV2. They will be ignored

cmccabe commented on metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java:
```diff
@@ -354,6 +357,25 @@ public ControllerResult<BrokerRegistrationReply> registerBroker(
+            if (set.size() != request.logDirs().size()) {
+                throw new InvalidRequestException("Duplicate directory ID in request");
+            }
+            for (BrokerRegistration registration : brokerRegistrations().values()) {
```
I think we should have a custom error code for this situation. InvalidRequestException seems too generic.

Also, we shouldn't be checking the previous registration for this specific broker ID... it would be quite normal to re-register broker ID X with the same set of directories it had last time we registered it.

cmccabe replied:
Another thing. This is `O(num_brokers * num_directories)` which isn't good for larger deployments. I think we should just have a `TimelineHashMap<Uuid, Integer>` that maps log dir UUID to the broker that has it. The main hassle is removing the mappings during broker unregistration, but that is rare, of course.

cmccabe replied:
> Also, we shouldn't be checking the previous registration for this specific broker ID... it would be quite normal to re-register broker ID X with the same set of directories it had last time we registered it

This might be what's causing the test failures btw

soarez replied:
@cmccabe:

>  InvalidRequestException seems too generic

I've introduced InvalidRegistrationException to handle these.

> Also, we shouldn't be checking the previous registration for this specific broker ID.

I forgot that. Thanks for pointing this out!

> we should just have a TimelineHashMap<Uuid, Integer> that maps log dir UUID to the broker that has it

Great suggestion. Added.

cmccabe commented on metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java:
```diff
@@ -192,7 +192,7 @@ private BrokerRegistration(
         directories = new ArrayList<>(directories);
         directories.sort(Uuid::compareTo);
-        this.directories = Collections.unmodifiableList(directories);
+        this.sortedDirectories = Collections.unmodifiableList(directories);
```
Seems like we should copy the list and sort it here to make sure that it actually is sorted. It should be a short list (a dozen or so max) so not a big performance issue.

rondagostino replied:
We do that in the prior 2 lines

cmccabe replied:
ah, sorry, that is my bad. Please ignore this then :)

cmccabe commented on metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java:
```diff
@@ -337,7 +341,7 @@ public boolean equals(Object o) {
             other.inControlledShutdown == inControlledShutdown &&
             other.isMigratingZkBroker == isMigratingZkBroker &&
-            other.directories.equals(directories);
+            other.sortedDirectories.equals(sortedDirectories);
```
This is another reason to go with sorted -- `list.equals` won't do the right thing otherwise.

cmccabe commented on metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java:
```diff
@@ -272,8 +274,17 @@ public void testRegistrationWithIncorrectClusterId() {
+        return Stream.of(
+                MetadataVersion.IBP_3_3_IV2,
+                MetadataVersion.IBP_3_3_IV3,
+                MetadataVersion.LATEST_PRODUCTION,
```
I feel like we should explicitly include `IBP_3_7_IV2` here rather than `LATEST_PRODUCTION`. After all, that is the first JBOD version so it's worth testing specifically, even after `LATEST_PRODUCTION` marches on.

soarez replied:
Replaced `LATEST_PRODUCTION` with `IBP_3_7_IV2`.

cmccabe commented:
Sorry, one more thing: I didn't understand the purpose of the HeartbeatManager changes. Seems like we should be able to keep this file the same now? Or if not, can we do them in a separate PR where we have some rationale

soarez committed:
KAFKA-15361: Process and persist dir info with broker registration

Controllers should process and persist directory information from the broker registration request

soarez commented:
@cmccabe:

> I didn't understand the purpose of the HeartbeatManager changes (now that we've agreed to keep Directory out of HBM). Seems like we should be able to keep this file the same now? Or if not, can we change it in a separate PR where we have some rationale?

The changes aren't necessary, but they are an improvement. I've moved them to a separate PR: #14934

rondagostino commented:
[Still 110 failing tests](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/14/tests), which seems like a lot, especially since https://github.com/apache/kafka/pull/14934/ only shows [18 failing tests](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14934/1/tests/).

I've [restarted the build](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/15/pipeline) in the hope that we see many fewer test failures.

soarez commented:
@rondagostino I'm trying to figure the test failures out, I've not yet been able to reproduce any failures locally.

rondagostino commented:
[Latest builds for JDKs 8 and 21 failed](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/16/pipeline/).  Restarted the build: https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14838/17/pipeline

So frustrating.

cmccabe commented:
@rondagostino : I think we can merge this now because I don't think it's adding more failures. I do think we need to do another sweep looking for people calling `Exit.exit`, but that doesn't need to block here.

lucasbru commented:
There are some tests that seem to have gotten a lot worse since this PR was merged, as far as I can tell, e.g. `ZkMigrationIntegrationTest`.

https://ge.apache.org/scans/tests?search.rootProjectNames=kafka&search.timeZoneId=Europe%2FBerlin&tests.container=kafka.zk.ZkMigrationIntegrationTest&tests.test=testMigrateTopicDeletions(ClusterInstance)%5B6%5D

dajac commented:
@soarez Could you take a look?

soarez commented:
@dajac: I'm having a look. `ZkMigrationIntegrationTest.testMigrateTopicDeletions` and `ZkMigrationIntegrationTest.testDualWrite` are consistently failing for `IBP_3_7_IV2` or greater, suggesting this is indeed JBOD related.

soarez commented:
@dajac, @lucasbru: Please take a look at #14976
